## Confirm CSV file exist and is not empty

In [1]:
from pathlib import Path

In [2]:
file_path = Path("../data/raw/nba_2025_26_player_per_game.csv")

In [3]:
print("Exists:", file_path.exists())

Exists: True


In [4]:
print("Size:", file_path.stat().st_size, "bytes")

Size: 104853 bytes


## Test whether pandas can parse the CSV file. Ensure no ParserError


In [5]:
import pandas as pd

In [6]:
df = pd.read_csv(file_path)

## 1. Initial Inspection


In [7]:
df.shape

(734, 32)

In [8]:
df.head()

,Rk,Player,Age,Team,Pos,G,GS,MP,FG,FGA,...,DRB,TRB,AST,STL,BLK,TOV,PF,PTS,Awards,Player-additional
0,1.0,Luka Dončić,26.0,LAL,PG,64.0,64.0,35.8,10.8,22.8,...,7.1,7.7,8.3,1.6,0.5,4.0,2.4,33.5,MVP-4CPOY-8ASNBA1,doncilu01
1,2.0,Shai Gilgeous-Alexander,27.0,OKC,PG,68.0,68.0,33.2,10.8,19.4,...,3.7,4.3,6.6,1.4,0.8,2.2,2.0,31.1,MVP-1CPOY-1ASNBA1,gilgesh01
2,3.0,Anthony Edwards,24.0,MIN,SG,61.0,60.0,35.0,9.9,20.2,...,4.4,5.0,3.7,1.4,0.8,2.9,1.9,28.8,CPOY-3AS,edwaran01
3,4.0,Jaylen Brown,29.0,BOS,SF,71.0,71.0,34.4,10.4,21.7,...,5.8,6.9,5.1,1.0,0.4,3.6,2.7,28.7,MVP-6ASNBA2,brownja02
4,5.0,Tyrese Maxey,25.0,PHI,PG,70.0,70.0,38.0,9.9,21.4,...,3.8,4.1,6.6,1.9,0.8,2.4,2.2,28.3,CPOY-7ASNBA3,maxeyty01


In [9]:
df.tail()

,Rk,Player,Age,Team,Pos,G,GS,MP,FG,FGA,...,DRB,TRB,AST,STL,BLK,TOV,PF,PTS,Awards,Player-additional
729,579.0,Darius Brown II,26.0,CLE,SG,1.0,0.0,3.0,0.0,1.0,...,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,NaN,brownda04
730,580.0,Noa Essengue,19.0,CHI,PF,2.0,0.0,3.0,0.0,1.5,...,0.0,0.0,0.0,0.5,0.0,0.0,0.5,0.0,NaN,essenno01
731,581.0,Tosan Evbuomwan,24.0,NYK,SF,5.0,0.0,1.6,0.0,0.2,...,0.2,0.4,0.0,0.0,0.0,0.0,0.2,0.0,NaN,evbuoto01
732,582.0,Chucky Hepburn,22.0,TOR,PG,2.0,0.0,6.5,0.0,3.0,...,0.5,0.5,1.0,0.5,0.0,0.5,1.0,0.0,NaN,hepbuch01
733,NaN,League Average,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,-9999


In [10]:
import csv
from collections import Counter

with open(file_path, encoding="utf-8-sig", newline="") as file:
    column_counts = Counter(
        len(row) for row in csv.reader(file) if row
    )

column_counts

Counter({32: 735})

In [11]:
df[df["Player"] == "Player"]

,Rk,Player,Age,Team,Pos,G,GS,MP,FG,FGA,...,DRB,TRB,AST,STL,BLK,TOV,PF,PTS,Awards,Player-additional


In [12]:
df[
    df["Player"].str.contains("Don", case=False, na=False)
][["Player", "Team", "G", "MP", "PTS"]]

,Player,Team,G,MP,PTS
0,Luka Dončić,LAL,64.0,35.8,33.5
6,Donovan Mitchell,CLE,70.0,33.5,27.9
30,Brandon Ingram,TOR,77.0,33.8,21.5
37,Shaedon Sharpe,POR,50.0,29.4,20.8
43,Brandon Miller,CHO,65.0,30.3,20.2
103,Aaron Gordon,DEN,36.0,27.9,16.2
167,Keldon Johnson,SAS,82.0,23.3,13.2
170,Brandon Williams,DAL,66.0,22.2,13.0
184,Donte DiVincenzo,MIN,82.0,30.4,12.2
188,Donovan Clingan,POR,77.0,27.2,12.1


In [13]:
print("Total rows:", len(df))

Total rows: 734


In [14]:
print("Unique ranks:", df["Rk"].nunique())

Unique ranks: 582


In [15]:
print("Unique player names:", df["Player"].nunique())

Unique player names: 583


In [16]:
df["Player"].value_counts().head(10)

Player
Charles Bassey         5
Kyle Anderson          4
RayJ Dennis            4
Patrick Baldwin Jr.    4
Tyus Jones             4
Christian Koloko       4
James Harden           3
Jaren Jackson Jr.      3
Darius Garland         3
CJ McCollum            3
Name: count, dtype: int64

In [17]:
duplicate_players = df[
    df.duplicated(subset="Player", keep=False)
]

duplicate_players[
    ["Rk", "Player", "Team", "G", "MP", "PTS"]
].sort_values(["Player", "Team"]).head(30)

,Rk,Player,Team,G,MP,PTS
601,475.0,AJ Johnson,2TM,48.0,9.5,3.3
603,475.0,AJ Johnson,DAL,23.0,10.4,3.9
602,475.0,AJ Johnson,WAS,25.0,8.6,2.8
610,482.0,Amir Coffey,2TM,46.0,10.7,3.2
611,482.0,Amir Coffey,MIL,30.0,8.8,2.4
612,482.0,Amir Coffey,PHO,16.0,14.1,4.8
130,109.0,Anfernee Simons,2TM,55.0,24.9,14.3
131,109.0,Anfernee Simons,BOS,49.0,24.5,14.2
132,109.0,Anfernee Simons,CHI,6.0,28.3,15.2
122,103.0,Ayo Dosunmu,2TM,69.0,27.3,14.8


In [18]:
df[
    ["Player", "Player-additional"]
].head(10)

,Player,Player-additional
0,Luka Dončić,doncilu01
1,Shai Gilgeous-Alexander,gilgesh01
2,Anthony Edwards,edwaran01
3,Jaylen Brown,brownja02
4,Tyrese Maxey,maxeyty01
5,Kawhi Leonard,leonaka01
6,Donovan Mitchell,mitchdo01
7,Nikola Jokić,jokicni01
8,Giannis Antetokounmpo,antetgi01
9,Joel Embiid,embiijo01


In [19]:
df["Player-additional"].nunique()

583

In [20]:
df["Player-additional"].isna().sum()

np.int64(0)

### Initial CSV integrity findings

- The CSV contains 734 data rows and 32 columns.
- All non-empty CSV rows contain the same number of fields.
- The additional `Player-additional` field appears to be a source identifier.
- The pandas index represents row position, not player rank.
- There are more rows than ranked players because some players have aggregate and team-specific rows after changing teams.
- The dataset also includes a non-player League Average row.
- The file appears structurally intact, but its content and statistics still require further validation.

### Inspect data types and missing values


In [21]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 734 entries, 0 to 733
Data columns (total 32 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 733 non-null    float64
 1   Player             734 non-null    str    
 2   Age                733 non-null    float64
 3   Team               733 non-null    str    
 4   Pos                733 non-null    str    
 5   G                  733 non-null    float64
 6   GS                 733 non-null    float64
 7   MP                 733 non-null    float64
 8   FG                 733 non-null    float64
 9   FGA                733 non-null    float64
 10  FG%                731 non-null    float64
 11  3P                 733 non-null    float64
 12  3PA                733 non-null    float64
 13  3P%                683 non-null    float64
 14  2P                 733 non-null    float64
 15  2PA                733 non-null    float64
 16  2P%                727 non-null    fl

#### All rows are appropriately in strings or float.

In [22]:
df.dtypes

Rk                   float64
Player                   str
Age                  float64
Team                     str
Pos                      str
G                    float64
GS                   float64
MP                   float64
FG                   float64
FGA                  float64
FG%                  float64
3P                   float64
3PA                  float64
3P%                  float64
2P                   float64
2PA                  float64
2P%                  float64
eFG%                 float64
FT                   float64
FTA                  float64
FT%                  float64
ORB                  float64
DRB                  float64
TRB                  float64
AST                  float64
STL                  float64
BLK                  float64
TOV                  float64
PF                   float64
PTS                  float64
Awards                   str
Player-additional        str
dtype: object